feat: Implement manualSessionHandling and connect UI with analytics operations - WPB-10299

agisilaos opened the pull request:
<!--do not remove this marker, its needed to replace info when ticket title is updated -->
<!--jira-description-action-hidden-marker-start-->

<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-10299" title="WPB-10299" target="_blank"><img alt="Task" src="https://wearezeta.atlassian.net/rest/api/2/universal_avatar/view/type/issuetype/avatar/10818?size=medium" />WPB-10299</a>  [iOS] Check if manualSessionHandling needs to be enabled for Countly Configuration
  </td></table>
  <br />
 

<!--jira-description-action-hidden-marker-end-->
<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove the jira markers to link tickets automatically -->


With this PR we implement manualSessionHandling and we connect our new analytics implementation with the UI (meaning enabling/disabling) More specifically we introduce two new use cases: 

- `EnableAnalyticsSharingUseCase`
- `DisableAnalyticsSharingUseCase`

In addition to that `TrackingManager` is not a singleton anymore. 

### TODO:
- [x] Push the changes for the alert
- [x] Add tests for the use cases 

---

### Checklist

- [x] Title contains a reference JIRA issue number like `[WPB-XXX]`.
- [x] Description is filled and free of optional paragraphs.
- [x] Adds/updates automated tests.

---

agisilaos committed:
fix: Use objectForkey for safety

agisilaos committed:
Introduce a new useCase

agisilaos committed:
Introduce disable to AnalyticsManagerProtocol

agisilaos committed:
Implement disable method

agisilaos committed:
Make AnalyticsService Protocol public

agisilaos committed:
Introduce making the use case in user session

agisilaos committed:
Add documentation to the method

agisilaos committed:
Introduce the manager to the ZMUserSession

agisilaos committed:
Fix Build errors

agisilaos committed:
Merge branch 'epic/analytics' into feat/connect-ui-with-analytics

agisilaos committed:
feat: Make TrackingManager not a singleton

agisilaos committed:
Remove calls to Analytics.shared

agisilaos committed:
Set manualSessionHandling to true

agisilaos committed:
Remove makeDisableAnalyticsSharingUseCase

agisilaos committed:
Introduce the SessionManager to the TrackingManager

agisilaos committed:
Introduce enableTracking method

Cleanup AnalyticsManager

agisilaos committed:
Groom

agisilaos committed:
Groom DisableAnalyticsSharingUseCase

agisilaos committed:
Introduce Enable AnalyticsSharingUseCase and groom

agisilaos committed:
Groom implementation

agisilaos committed:
disableAnalyticsSharing = true initially

agisilaos committed:
Merge branch 'epic/analytics' into feat/connect-ui-with-analytics

agisilaos committed:
Revert "disableAnalyticsSharing = true initially"

This reverts commit 6bf5e500d51682cd169f73a0423a7be8c3360d37.

agisilaos committed:
wip alert

agisilaos committed:
Revert "wip alert"

This reverts commit 68048fd0549e3e86fd2e7ce05712ab1ad6217b8e.

agisilaos committed:
Rename file name for usecase

agisilaos committed:
Reapply "disableAnalyticsSharing = true initially"

This reverts commit 0669b1bda9e30393e49c0e9a2dafb959ebb48a8d.

agisilaos committed:
Fix SwiftLint warnings

agisilaos committed:
Introduce new strings for alerts

agisilaos committed:
Introduce new alert

agisilaos committed:
Improve the name of the property

agisilaos committed:
Introduce mocks for use cases

johnxnguyen commented on WireAnalytics/Sources/WireAnalytics/AnalyticsManager.swift:
```diff
@@ -53,15 +57,42 @@ public struct AnalyticsManager: AnalyticsManagerProtocol {
+
+    /// Disables tracking by ending the current session and clearing user data.
+    public func disableTracking() {
+        clearUserData()
+        analyticsService.endSession()
```
It might be better to end the session first, then clear the user data.

agisilaos replied:
Agreed, fixed it here: https://github.com/wireapp/wire-ios/commit/d5ca76c32819a365b4035eb77982d7323c6b8a7f

johnxnguyen commented on WireAnalytics/Sources/WireAnalytics/AnalyticsService.swift:
```diff
@@ -20,7 +20,7 @@ import Foundation
 // sourcery: AutoMockable
 /// Protocol defining the requirements for an analytics service.
-protocol AnalyticsService {
+public protocol AnalyticsService {
```
Why does this need to be public?

johnxnguyen commented on wire-ios-sync-engine/Source/SessionManager/SessionManager+AnalyticsUseCases.swift:
```diff
@@ -0,0 +1,56 @@
+
+extension SessionManager {
+
+    enum SessionManagerError: Error {
```
The name of this error suggests it's for general errors within the session manager, but it's all to do with analytics. I suggest a more accurate name.

agisilaos replied:
You're absolutely right the name doesn't reflect the for those types of errors. Updated it here: https://github.com/wireapp/wire-ios/commit/5690c91fecccb9747ee0e35b7fd6f2ee9012d2de

johnxnguyen commented on wire-ios-sync-engine/Source/SessionManager/SessionManager+AnalyticsUseCases.swift:
```diff
@@ -0,0 +1,56 @@
+
+    public func makeDisableAnalyticsSharingUseCase() throws -> DisableAnalyticsSharingUseCaseProtocol {
+        guard let analyticsManager else {
+            throw SessionManagerError.missingAnalyticsManager
```
Suggestion: `analyticsNotAvailable`

agisilaos replied:
Done: https://github.com/wireapp/wire-ios/commit/5690c91fecccb9747ee0e35b7fd6f2ee9012d2de

johnxnguyen commented on wire-ios-sync-engine/Source/SessionManager/SessionManager.swift:
```diff
@@ -891,11 +891,19 @@ public final class SessionManager: NSObject, SessionManagerType {
+        activeUserSession.analyticsSession = analyticsSession
+    }
+
+    func getUserAnalyticsProfileForActiveUserSession() -> AnalyticsUserProfile? {
```
I suggest to make this an extension on the user session, or a function that accepts the user session as an argument. This would avoid unwrapping the user session twice.

agisilaos replied:
I decided to go with the function route: https://github.com/wireapp/wire-ios/commit/bec25454f8ca7ddc9ac4525626707f404f387658

johnxnguyen commented on wire-ios-sync-engine/Source/Use cases/DisableAnalyticsSharingUseCase.swift:
```diff
@@ -0,0 +1,52 @@
+public struct DisableAnalyticsSharingUseCase: DisableAnalyticsSharingUseCaseProtocol {
+
+    /// The analytics manager responsible for handling tracking operations.
+    private let analyticsManager: AnalyticsManagerProtocol?
```
Why is it optional?

agisilaos replied:
I don't think it needs to be. Only in the case where we have analytics disabled for specific builds... 🤔 Not sure. Removed optionality here: https://github.com/wireapp/wire-ios/commit/d239e12ad69ed903c7dc4527af5b2b1d4d00fc61

johnxnguyen commented on wire-ios-sync-engine/Source/Use cases/DisableAnalyticsSharingUseCase.swift:
```diff
@@ -0,0 +1,52 @@
+
+// sourcery: AutoMockable
+/// Protocol defining the interface for disabling analytics sharing.
+public protocol DisableAnalyticsSharingUseCaseProtocol {
```
Suggestion: `DisableAnalyticsUseCaseProtocol`

agisilaos replied:
Updated here: https://github.com/wireapp/wire-ios/commit/3ce973cac5f43ee041a4948f91a3257fd15cb357

& here https://github.com/wireapp/wire-ios/commit/7f9d8bc8783614c5e0b0a32abc12b84fdce74b6e

johnxnguyen commented on wire-ios-sync-engine/Source/Use cases/EnableAnalyticsSharingUseCase.swift:
```diff
@@ -0,0 +1,75 @@
+
+// sourcery: AutoMockable
+/// Protocol defining the interface for enabling analytics sharing.
+public protocol EnableAnalyticsSharingUseCaseProtocol {
```
Suggestion: `EnableAnalyticsUseCaseProtocol`

agisilaos replied:
Updated here: https://github.com/wireapp/wire-ios/commit/3ce973cac5f43ee041a4948f91a3257fd15cb357

& here https://github.com/wireapp/wire-ios/commit/7f9d8bc8783614c5e0b0a32abc12b84fdce74b6e

johnxnguyen commented on wire-ios-sync-engine/Source/Use cases/EnableAnalyticsSharingUseCase.swift:
```diff
@@ -0,0 +1,75 @@
+    // MARK: - Properties
+
+    /// The analytics manager responsible for handling tracking operations.
+    private let analyticsManager: AnalyticsManagerProtocol?
```
Why is this optional?

agisilaos replied:
I don't think it needs to be. Only in the case where we have analytics disabled for specific builds... 🤔 Not sure. Removed optionality here: https://github.com/wireapp/wire-ios/commit/d239e12ad69ed903c7dc4527af5b2b1d4d00fc61

johnxnguyen commented on wire-ios-sync-engine/Source/UserSession/ZMUserSession/ZMUserSession.swift:
```diff
@@ -81,6 +81,8 @@ public final class ZMUserSession: NSObject {
 
     var analyticsSession: (any AnalyticsSessionProtocol)?
 
+    var analyticsManager: (any AnalyticsManagerProtocol)?
```
What is this needed?

agisilaos replied:
It is not needed anymore. Good catch: https://github.com/wireapp/wire-ios/commit/92cd5919408c490af8ff75cc74235101e008468b

johnxnguyen commented on wire-ios/Wire-iOS/Resources/Base.lproj/Localizable.strings:
```diff
@@ -1160,7 +1160,11 @@
 
-
+"self.settings.privacy_analytics.alert.title" = "Consent to the Analysis of Your User Behaviour";
+"self.settings.privacy_analytics.alert.message" = "With your consent, we use technologies to analyse the use of Wire Messenger in order to improve and develop it based on user needs. Information about your use of Wire Messenger, such as whether and when an error occurs or a particular function is used (so-called events), and other technical information (e.g., IP address, app version, device type, ...
```
> The ID is deleted from your device after [28] days after the last activity at the latest.

Where does this happen?

agisilaos replied:
Provided an answer for you 1:1.

johnxnguyen commented on wire-ios/Wire-iOS/Sources/Analytics/TrackingManager.swift:
```diff
@@ -22,24 +22,47 @@ import WireCommonComponents
+                    enableUseCase.invoke()
+                }
+            } catch {
+                print("Failed to toggle analytics sharing: \(error)")
```
This should be a log, and in case it fails, we need to reset the setting.

agisilaos replied:
Correct! Fixed it here: https://github.com/wireapp/wire-ios/commit/1bd1e64892df204e1cfd17260a6b1ba72a15e3a2

johnxnguyen commented on wire-ios/Wire-iOS/Sources/AppRootRouter.swift:
```diff
@@ -421,21 +423,22 @@ extension AppRootRouter {
 
-        TrackingManager.shared.disableAnalyticsSharing = false
-        Analytics.shared.provider?.selfUser = selfUser
+        trackingManager.disableAnalyticsSharing = true
```
Won't this override the user's stored preference every time the app is launched?

agisilaos replied:
Yep! Fixed it here: https://github.com/wireapp/wire-ios/commit/bd2750fe7733ccabe8628e32d9eb1806e67cc176

johnxnguyen commented on wire-ios/Wire-iOS/Sources/Components/Settings/SettingsPropertyFactory.swift:
```diff
@@ -231,12 +235,27 @@ final class SettingsPropertyFactory {
+                            self.showAnalyticsConsentAlert { accepted in
+                                if accepted {
+                                    tracking.disableAnalyticsSharing = newValue
+                                    AVSFlowManager.getInstance()?.setEnableMetrics(!newValue)
```
Isn't this part of the tracking manager? Why do it here as well?

agisilaos replied:
Moved it to the trackingManager: https://github.com/wireapp/wire-ios/commit/508f597ce80e6ab2a0a16c8dfb8b03f4c045901c

johnxnguyen commented on wire-ios/Wire-iOS/Sources/LaunchSequenceOperation.swift:
```diff
@@ -86,7 +86,7 @@ final class TrackingOperation: LaunchSequenceOperation {
 
         AnalyticsProviderFactory.shared.useConsoleAnalytics = containsConsoleAnalytics
-        Analytics.shared = Analytics(optedOut: TrackingManager.shared.disableAnalyticsSharing)
+       // Analytics.shared = Analytics(optedOut: TrackingManager.shared.disableAnalyticsSharing)
```
To remove?

agisilaos replied:
Yep, removed it here: https://github.com/wireapp/wire-ios/commit/66eb745653a5cab5098e1264ebbb6402d69820dd

johnxnguyen commented on wire-ios/Wire-iOS/Sources/UserInterface/Settings/CellDescriptors/SettingsCellDescriptorFactory+Account.swift:
```diff
@@ -384,7 +384,7 @@ extension SettingsCellDescriptorFactory {
 
     func signOutElement() -> SettingsCellDescriptorType {
-        return SettingsSignOutCellDescriptor()
+        return SettingsSignOutCellDescriptor(trackingManager: trackingManager)
```
Why does it need the tracking manager? It seems unrelated no?

agisilaos replied:
It was needed because we were setting the disableAnalyticsSharing property inside the `private func logout(password: String? = nil)` since I just removed that it's not needed anymore. I removed the trackingManager from the descriptor as well. https://github.com/wireapp/wire-ios/commit/a8288ec6ca5c59eacd5afbfa2b32558f2ac9deac

johnxnguyen commented on wire-ios/Wire-iOS/Sources/UserInterface/Settings/CellDescriptors/SettingsCellDescriptorFactory+Developer.swift:
```diff
@@ -98,7 +98,7 @@ extension SettingsCellDescriptorFactory {
         )
 
-        if !Analytics.shared.isOptedOut && !TrackingManager.shared.disableAnalyticsSharing {
+        if !trackingManager.disableAnalyticsSharing {
```
I wonder what role the `Analytics.shared.isOptedOut` played and how it's different to `disableAnalyticsSharing`. Do you know?

agisilaos replied:
`isOptedOut` is not used anymore and from what I can found here's how the getter and the setter is looking like:

```swift
import Foundation
import WireDataModel

extension Analytics {

    /// Opt the user out of sending analytics data
    var isOptedOut: Bool {
        get {
            guard let provider else { return true }
            return provider.isOptedOut
        }

        set {
            if newValue && (provider?.isOptedOut ?? false) {
                return
            }

            if newValue {
                provider?.flush {
                    self.provider?.isOptedOut = newValue
                    self.provider = nil
                }
            } else {
                provider = AnalyticsProviderFactory.shared.analyticsProvider()

                if let user = SelfUser.provider?.providedSelfUser {
                    selfUser = user
                } else {
                    assertionFailure("expected available 'user'!")
                    selfUser = nil
                }
            }
        }
    }
}
```

johnxnguyen commented on wire-ios/Wire-iOS/Sources/UserInterface/Settings/CellDescriptors/SettingsSignOutCellDescriptor.swift:
```diff
@@ -54,7 +56,7 @@ final class SettingsSignOutCellDescriptor: SettingsExternalScreenCellDescriptor
             ZMUserSession.shared()?.logout(credentials: UserEmailCredentials(email: "", password: password ?? "")) { [weak topMostViewController] result in
                 Task { @MainActor in self.activityIndicator.stop() }
-                TrackingManager.shared.disableAnalyticsSharing = false
+                self.trackingManager.disableAnalyticsSharing = false
```
Why does logging out change if analytics is enabled?

agisilaos replied:
It was there in the previous implementation but for the new one we don't have anything that explicitly says we need to do anything regarding analytics during log out. And because of that I removed the tracking manager entirely from the descriptor: https://github.com/wireapp/wire-ios/commit/a8288ec6ca5c59eacd5afbfa2b32558f2ac9deac

johnxnguyen reviewed (changes requested):
Are you planning to add tests?

agisilaos committed:
Introduce DisableAnalyticsSharingUseCaseTests

agisilaos committed:
Fix: SwiftLint Warnings

agisilaos committed:
Make analyticsManager nonOptional

agisilaos committed:
Remove not needed property

agisilaos committed:
Delete commented out line

agisilaos committed:
Remove TrackingManager from Descriptor

agisilaos committed:
Make AnalyticsService non public

agisilaos committed:
Provide a more suitable name for the error

agisilaos committed:
Update the names of the Protocols

agisilaos committed:
Better names for use cases

agisilaos committed:
Don't override user's choice

agisilaos committed:
First endSession then clearData

agisilaos committed:
Log error instead of print and reset the setting

agisilaos committed:
Use a type for safety

The stringly-typed keys is error prone, maybe we can create a CaseIterable enum to make sure we remove all values.

agisilaos commented on WireAnalytics/Sources/WireAnalytics/AnalyticsManager.swift:
```diff
@@ -53,15 +57,42 @@ public struct AnalyticsManager: AnalyticsManagerProtocol {
+        for key in ["team_team_id", "team_user_type", "team_team_size"] {
+            analyticsService.setUserValue(nil, forKey: key)
+        }
+    }
```
Not really sure about the `CLYTemporaryDeviceID` but regarding having a more concrete type for safety. I implemented it here and added documentation: https://github.com/wireapp/wire-ios/commit/fc2ac7c9ab8f7fe91ba20a83a08489a0e522918e

agisilaos committed:
Avoid unwrapping userSession twice

samwyndham commented on wire-ios-sync-engine/Source/SessionManager/SessionManager.swift:
```diff
@@ -310,7 +310,7 @@ public final class SessionManager: NSObject, SessionManagerType {
     private let minTLSVersion: String?
 
-    private var analyticsManager: (any AnalyticsManagerProtocol)?
+    var analyticsManager: (any AnalyticsManagerProtocol)?
```
I think this can be `private(set) var analyticsManager: (any AnalyticsManagerProtocol)?` as it is not `set` from outside

agisilaos replied:
Done: https://github.com/wireapp/wire-ios/commit/3946e2aeb8cd51f158b993d8cb73c180c15feb38

samwyndham commented on wire-ios/Wire-iOS/Sources/Components/Settings/SettingsPropertyFactory+UIAlertController.swift:
```diff
@@ -0,0 +1,54 @@
+            alertController.addAction(UIAlertAction(title: action.title, style: action.style, handler: action.handler))
+        }
+
+        UIApplication.shared.topmostViewController(onlyFullScreen: false)?
```
If there is no `topmostViewController` this would be a serious issue. I think you should guard let for it and log an error if it is not there.

agisilaos replied:
Done: https://github.com/wireapp/wire-ios/commit/cd5db3f9625e59c227722780a71e1e219e12d4aa

samwyndham commented on wire-ios-sync-engine/Source/SessionManager/SessionManager.swift:
```diff
@@ -888,24 +888,29 @@ public final class SessionManager: NSObject, SessionManagerType {
     }
 
-    private func switchAnalyticsUser() {
+    private func switchAnalyticsUser(for userSession: ZMUserSession) {
```
I think this should be `to` instead of `for`. Small thing but I think otherwise it could be confusing.

samwyndham commented on wire-ios-sync-engine/Source/SessionManager/SessionManager.swift:
```diff
@@ -888,24 +888,29 @@ public final class SessionManager: NSObject, SessionManagerType {
-            let activeUserSession
+            let userProfile = getUserAnalyticsProfile(for: userSession)
         else {
             return
```
Is this a valid state? 

If not I think you should log the error and do an `assertionFailure()` to surface this issue, especially since there are a lot of possible ways `getUserAnalyticsProfile(for:)` can return nil

agisilaos replied:
It is a valid state in some cases analytics are disabled etc.

samwyndham commented on wire-ios/Wire-iOS/Sources/Analytics/TrackingManager.swift:
```diff
@@ -22,24 +22,51 @@ import WireCommonComponents
+        super.init()
+    }
 
     var disableAnalyticsSharing: Bool {
```
I think that this property should be tested if it is not already being tested.

agisilaos replied:
Out of the scope of this PR but definitely on my mind to do so.

agisilaos committed:
Merge branch 'epic/analytics' into feat/connect-ui-with-analytics

agisilaos committed:
Add a warning to a method's documentation

agisilaos commented on WireAnalytics/Sources/WireAnalytics/AnalyticsManagerProtocol.swift:
```diff
@@ -23,4 +23,14 @@ public protocol AnalyticsManagerProtocol {
+    func disableTracking()
+
+    /// Enables analytics tracking and reporting.
+    /// This method should be used to re-enable analytics tracking after it has been disabled. Ensure that any necessary state or user context is correctly restored before resuming tracking.
```
Done: https://github.com/wireapp/wire-ios/commit/211348f6122f08a19b22bdeccc6b77ae3fd5120e

agisilaos committed:
Use a guard let for topViewController and log it

agisilaos committed:
Rename method for better clarity

agisilaos committed:
Update AutoMockable.generated.swift

agisilaos committed:
mark property with private(set)

agisilaos committed:
Remove not needed comments

agisilaos commented on wire-ios-sync-engine/Source/Use cases/EnableAnalyticsUseCase.swift:
```diff
@@ -0,0 +1,75 @@
+
+    // MARK: - Properties
+
+    /// The analytics manager responsible for handling tracking operations.
```
Done: https://github.com/wireapp/wire-ios/commit/fe797489d4c9684ecb31b7c52d4168fe7c914477

agisilaos committed:
Fix: SwiftLint warning

agisilaos committed:
Delete Analytics+OptOut.swift

agisilaos committed:
Set trackingManager to nil in tests

agisilaos committed:
Remove `trackingManager` from `SettingsCellDescriptorFactory`

It's not needed

agisilaos committed:
Set `trackingManager` to nil

agisilaos committed:
Make `TrackingManager` optional

agisilaos committed:
Delete not needed code

agisilaos committed:
Update Tests by setting `trackingManager` to nil

agisilaos committed:
Delete unused code

agisilaos committed:
Fix SwiftLint Warnings

agisilaos committed:
Address code review comment

agisilaos committed:
Move `showAnalyticsConsentAlert` to `TrackingManager`

agisilaos committed:
Merge branch 'epic/analytics' into feat/connect-ui-with-analytics

johnxnguyen commented on WireAnalytics/Sources/WireAnalytics/AnalyticsManager.swift:
```diff
@@ -84,14 +84,22 @@ public struct AnalyticsManager: AnalyticsManagerProtocol {
-        analyticsService.setUserValue(userProfile.teamInfo?.role, forKey: "team_user_type")
-        analyticsService.setUserValue(userProfile.teamInfo.map { String($0.size.logRound()) }, forKey: "team_team_size")
+
+        for key in AnalyticsUserKey.allCases {
+            switch key {
+            case .teamTeamID:
+                analyticsService.setUserValue(userProfile.teamInfo?.id, forKey: key.rawValue)
+            case .teamUserType:
+                analyticsService.setUserValue(userProfile.teamInfo?.role, forKey: key.rawValue)
+            case .teamTeamSize:
+                analyticsService.setUserValue(userProfile.teamInfo.map { String($0.size.logRound()) }, forKey: key.rawValue)
+            }
+        }
```
I find this a little confusing, I suggest to be simple:

```swift
analyticsService.setUserValue(userProfile.teamInfo?.id, forKey: AnalyticsUserKey.teamID.rawValue)
analyticsService.setUserValue(userProfile.teamInfo?.role, forKey: AnalyticsUserKey.teamRole.rawValue)
...
```

agisilaos replied:
Done: https://github.com/wireapp/wire-ios/commit/e4ab19dde01778aa71df214d8c89191ca7c343e6

johnxnguyen commented on WireAnalytics/Sources/WireAnalytics/AnalyticsManagerProtocol.swift:
```diff
@@ -28,9 +28,11 @@ public protocol AnalyticsManagerProtocol {
     /// - Parameter userProfile: The profile of the user to enable analytics for.
     /// - Returns: A new analytics session for the user.
+    ///
+    /// - Warning: Ensure that any necessary state or user context is correctly restored before resuming tracking.
```
Is this warning relevant for the caller of the method?

agisilaos replied:
Probably not. I removed it for now. https://github.com/wireapp/wire-ios/commit/fc11fef203a482afe637e069c08381b8d64e956b

johnxnguyen commented on WireAnalytics/Sources/WireAnalytics/AnalyticsUserKey.swift:
```diff
@@ -0,0 +1,33 @@
+enum AnalyticsUserKey: String, CaseIterable {
+
+    /// The key for the team ID.
+    case teamTeamID = "team_team_id"
+
+    /// The key for the user type within the team.
+    case teamUserType = "team_user_type"
+
+    /// The key for the size of the team.
+    case teamTeamSize = "team_team_size"
```
These names are confusing, I suggest `teamID`, `teamRole`, `teamSize`.

agisilaos replied:
Done: https://github.com/wireapp/wire-ios/commit/70ce8a8bd20ff42f5d9f75590180d1f90dbf6dbf

johnxnguyen commented on WireAnalytics/Sources/WireAnalyticsSupport/MockAnalyticsManagerProtocol.swift:
Is there a reason why this mocked manually and not with sourcery?

agisilaos replied:
`any` doesn't work well with Sourcery so I needed to create the mock manually.

johnxnguyen commented on wire-ios-sync-engine/Source/SessionManager/SessionManager.swift:
```diff
@@ -891,27 +891,24 @@ public final class SessionManager: NSObject, SessionManagerType {
         // Set the analytics identifier if it's not present
         if selfUser.analyticsIdentifier == nil {
-            let idProvider = AnalyticsIdentifierProvider(selfUser: activeUserSession.selfUser)
+            let idProvider = AnalyticsIdentifierProvider(selfUser: userSession.selfUser)
```
It's confusing that you access the self user it two different ways. I suggest to reuse the self user stored at the beginning of the method.

agisilaos replied:
Done: https://github.com/wireapp/wire-ios/commit/edeedc9c3a03de80d18e046fbe7705f22943fe27

johnxnguyen commented on wire-ios/Wire-iOS/Sources/Analytics/TrackingManager/TrackingManager.swift:
```diff
@@ -48,21 +48,36 @@ final class TrackingManager: NSObject, TrackingInterface {
-                if newValue {
-                    let disableUseCase = try sessionManager.makeDisableAnalyticsSharingUseCase()
-                    disableUseCase.invoke()
+            self.showAnalyticsConsentAlert { userConsented in
```
I don't understand... every time we set the value then we should ask for consent?

agisilaos replied:
Fixed it here: https://github.com/wireapp/wire-ios/pull/1767/commits/4d057e135cfeef971c25594ac2e148f186b7e24f

johnxnguyen commented on wire-ios/Wire-iOS/Sources/Analytics/TrackingManager/TrackingManager.swift:
```diff
@@ -48,21 +48,36 @@ final class TrackingManager: NSObject, TrackingInterface {
         }
     }
 
+    private func updateAnalyticsSharing(_ disable: Bool) {
```
I suggest to give the argument a name because it's hard to know what the bool means from the call site. E.g `updateAnalyticsSharing(true)`, what does true mean without looking at the implementation?

agisilaos replied:
Done: https://github.com/wireapp/wire-ios/pull/1767/commits/4d057e135cfeef971c25594ac2e148f186b7e24f

johnxnguyen commented on wire-ios/Wire-iOS/Sources/Analytics/TrackingManager/TrackingManager.swift:
```diff
@@ -48,21 +48,36 @@ final class TrackingManager: NSObject, TrackingInterface {
+            WireLogger.analytics.error("Failed to toggle analytics sharing: \(error)")
+
+            // Reset the setting
+            ExtensionSettings.shared.disableAnalyticsSharing = !disable
+            AVSFlowManager.getInstance()?.setEnableMetrics(disable)
```
This is error prone because you're assuming that the value was previously the opposite of the new value but there's no guard that is actually the case. 

Instead, you could invoke the use case and only after a success update the these settings.

agisilaos replied:
Done: https://github.com/wireapp/wire-ios/pull/1767/commits/4d057e135cfeef971c25594ac2e148f186b7e24f

agisilaos committed:
Address code review comments

agisilaos committed:
Have a common way to get the selfUser

agisilaos committed:
Rename properties

agisilaos committed:
Improve readability of the method

agisilaos committed:
Remove warning

agisilaos committed:
chore: Groom
